### PR TITLE
Annotate System.Text.Encoding.CodePages for nullable reference types

### PR DESCRIPTION
**Pull request by stephentoub**
Contributes to https://github.com/dotnet/corefx/issues/40623
cc: @buyaa-n, @safern, @krwq, @tarekgh

### CODE REVIEW

File: src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs [hunk 1]
```diff
@@ -1498,8 +1498,7 @@ private ISO2022Modes CheckEscapeSequenceKR(byte[] bytes, int escapeCount)
                         }
 
                         // Stick it in decoder
-                        if (decoder != null)
```
Comment (tarekgh) [inline src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs]: >     if (decoder != null) [](start = 19, length = 25)

I am not sure if removing this is correct. if the method GetCharsCP52936 is always called with valid decoder, then we should make it as not nullable and clean the code which check decoder==null. If GetCharsCP52936 is called with nullable decoder, we'll need to keep this check then and just move the next few lines under the condition.

File: src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs [hunk 2]
```diff
@@ -1576,8 +1575,7 @@ private ISO2022Modes CheckEscapeSequenceKR(byte[] bytes, int escapeCount)
                             break;
                         }
 
-                        if (decoder != null)
```
Comment (tarekgh) [inline src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs]: >decoder [](start = 28, length = 7)

same

File: src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs]: @tarekgh, just before this there's this code:
```C#
if (decoder == null || ...)
{
    ...
    break;
}
```
How could `decoder` be `null` here?

Comment (tarekgh) [reply src/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs]: you are right. I didn't notice the break. so we are fine then. 

---
In reply to: [337629043](https://github.com/dotnet/corefx/pull/41991#discussion_r337629043) [](ancestors = 337629043)

**Review by tarekgh — APPROVED** [review]

File: src/System.Text.Encoding.CodePages/src/System/Text/DecoderNLS.cs [hunk 3]
```diff
@@ -37,13 +37,6 @@ internal DecoderNLS(EncodingNLS encoding)
             Reset();
         }
 
-        // This is used by our child deserializers
```
Comment (safern) [inline src/System.Text.Encoding.CodePages/src/System/Text/DecoderNLS.cs]: Isn't this needed?

Comment (stephentoub) [reply src/System.Text.Encoding.CodePages/src/System/Text/DecoderNLS.cs]: The type isn't serializable in .NET Core.

**Review by safern — APPROVED** [review]

**Pull request merged**